### ISSUE TITLE
Create an example that uses wait() 

### ISSUE THREAD

**Issue by Vaishaal** [pywren/pywren#67]
I had to read the source to figure out how to make my calls synchronous. Would be nice to have a good example that gets this use pattern down.

(Also: why isn't there a version of map that is completely synchronous?)

**Comment by ericmjonas** [comment]
https://github.com/pywren/examples/blob/master/hello_world/hello_world.ipynb @Vaishaal does this  look like what you were thinking?

**Comment by shivaram** [comment]
@ericmjonas i'm worried that is too complex - one round of future waits + one round of future.result. Given we'd like non-CS users, would it make sense to have `map_sync()` that does this internally ? Or will that give up something basic ? 

P.S: If we go down this route, I'd also consider changing `map` to be synchronous and `map_async` to return futures

**Comment by ericmjonas** [comment]
@shivaram I'm ok with this complexity, the interface I'm trying to roughly approximate is https://docs.python.org/3/library/concurrent.futures.html. I'd much rather have a `get_all_results` which calls the blocking wait and gets the results.

**Comment by shivaram** [comment]
Ah I see - got it. So the `wait` is supposed to be similar to the [Python's wait]( https://docs.python.org/3.3/library/concurrent.futures.html#concurrent.futures.wait) ?

@Vaishaal what are your thoughts about `get_all_results`

**Comment by ericmjonas** [comment]
That's the hope. I mean, I know we can't perfectly reproduce that interface -- for example, at the moment we can't cancel jobs, and in general I think it's a bit dangerous to try and duplicated shared-memory-concurrency APIs for distributed systems (the failure probabilities are radically different) but -- it's also a pretty stable interface and one that's known by a lot of people. Plus it feels "pythonic".

**Comment by ooq** [comment]
I also thought making map() synchronous and another `map_async` is more natural. Of course, this is probably because I'm much more familiar with the typical `threadpool` interface than `concurrent.futures`.

**Comment by ooq** [comment]
One major edge of `concurrent.futures` seems to be fine-grained access over each call. You can check their status, blocking to return result, cancel them individually, etc. The question seems: 1) if they are useful; 2) if we can support them in a meaningful way, e.g., can we cancel invocations?

**Comment by shivaram** [comment]
I think its good to have the option to do things like that in the future. Also the async feature can let us compose or do lazy execution etc. So I think building on top of futures API is useful but at the same time having a sync API that works out of the box for users has its own attraction

BTW @ericmjonas one thing is we could make it look like a one-liner in our examples, README etc.
`results = get_all_results(map(...))` ?

**Comment by ericmjonas** [comment]
We could make it look like a one-liner, but I think that's hiding some real underlying complexity. I think target users are familiar with job invocation systems like PBS or condor so the idea of job handles that you wait on isn't all _that_ foreign. 

I'd much rather add `simplemap` in the future and leave things the way they are right now -- we have a lot of code written, I've not heard that this is a point of pain from users yet,  and we're trying to get a release out the door.
